skfda/misc/rkhs_product.py

```diff
@@ -0,0 +1,256 @@
+        fdata2: Second FData object.
+        cov_function: Covariance function as a callable.
+
+    Returns:
+        Matrix of the RKHS inner product between paired samples of
+        fdata1 and fdata2.
+    """
+    raise NotImplementedError(
+        "RKHS inner product not implemented for the given types.",
+    )
+
+
+@rkhs_inner_product.register(FDataGrid, FDataGrid)
+def rkhs_inner_product_fdatagrid(
+    fdata1: FDataGrid,
+    fdata2: FDataGrid,
+    *,
+    cov_function: Callable[[NDArrayFloat, NDArrayFloat], NDArrayFloat],
+    cond: Optional[float] = None,
+) -> NDArrayFloat:
+    r"""RKHS inner product of two univariate FDataGrid objects.
+
+    This is the most general method for calculating the RKHS inner product
+    using a discretization of the domain. The RKHS inner product is calculated
+    as the inner product between the square root inverse of the covariance
+    operator applied to the functions.
+    Assuming the existence of such inverse, using the self-adjointness of the
+    covariance operator, the RKHS inner product can be calculated as:
+        \langle f, \mathcal{K}^{-1} g \rangle
+    When discretizing a common domain, this is equivalent to doing the matrix
+    product between the discretized functions and the inverse of the
+    covariance matrix.
+    In case of having different discretization grids, the left inverse of the
+    transposed covariace matrix is used instead.
+
+    If one of the FDataGrid terms consists of only one sample, it is repeated
+    to match the number of samples of the other term.
+
+    Args:
+        fdata1: First FDataGrid object.
+        fdata2: Second FDataGrid object.
+        cov_function: Covariance function as a callable. It is expected to
+            receive two arrays, s and t, and return the corresponding
+            covariance matrix.
+        cond: Cutoff for small singular values of the covariance matrix.
+            Default uses scipy default.
+
+    Returns:
+        Matrix of the RKHS inner product between paired samples of
+        fdata1 and fdata2.
+    """
+    # Check univariate and apply broadcasting
+    check_fdata_dimensions(fdata1, dim_domain=1, dim_codomain=1)
+    check_fdata_dimensions(fdata2, dim_domain=1, dim_codomain=1)
+    fdata1, fdata2 = _broadcast_samples(fdata1, fdata2)
+
+    data_matrix_1 = fdata1.data_matrix
+    data_matrix_2 = fdata2.data_matrix
+    grid_points_1 = fdata1.grid_points[0]
+    grid_points_2 = fdata2.grid_points[0]
+    cov_matrix_1_2 = cov_function(grid_points_1, grid_points_2)
+
+    # Calculate the inverse operator applied to fdata2
+    if np.array_equal(grid_points_1, grid_points_2):
+        inv_fd2_matrix = np.linalg.solve(
+            cov_matrix_1_2,
+            data_matrix_2,
+        )
+    else:
+        inv_fd2_matrix = np.asarray(
+            scipy.linalg.lstsq(
+                cov_matrix_1_2.T,
+                data_matrix_2[..., 0].T,
+                cond=cond,
+            )[0],
```

Comment:
I find it hard to replace this transpositions because the parameters must be of dimensions (M,N) and (M,K) respectively. Since `cov_matrix` has dimensions (grid_points_of_fdata1, grid_points_of_fdata2) and `data_matrix` has dimensions (n_samples, grid_points, 1), in order to match dimensions both transpositions are required.